#Trabalho prático # 5
##Introdução à Computação Numérica

###Tópico: Solução de Problemas de Valor Inicial (PVI)

---

Nome: Vitória de Souza Serafim



---

**Instruções:** 

**Preencha o notebook** com seus dados, seus códigos e **rode todos os códigos** 

**Inclua  um relatório, na parte final do notebook,** com suas conclusões, respostas, observações e comentários relacionados com os tópicos e questões abordadas nas tarefas.  

**A elaboração dos códigos pode ser realizada em duplas, mas o resto do trabalho é individual.**

*Se você fez o seu trabalho em dupla inclua essa informação no relatório. Além disso, se você  também recebeu ajuda de um(a) colega/amigo(a) ou consultou/usou informações da internet ou outras fontes, coloque referências à elas no seu relatório.*


**Entrega do trabalho:**

*Data de entrega: 03/janeiro/2023* 

---


## Tarefa #1:

 Desenvolver códigos, na forma de funções que implementem: 

a) O Método de Euler. 

b) O Método  de Euler aperfeiçoado (ou Método de Heun). 

c) O Método de Runge-Kutta clássico de quarta ordem.



In [2]:
### importação do numpy
import numpy as np

In [3]:
# Função correspondente ao item a) 
####################################
# Metodo de Euler para o PVI
####################################

def met_euler( f, t0, tf, x0, np ):
    
    # listas para armazenar a discretização do intervalo 
    # e as aproximações da solução
    t = [t0]
    x = [x0]
    tn = t0
    xn = x0
    
    # tamanho do passo no tempo
    dt = float(tf-t0)/np

    # laço para o avanço no tempo
    for n in range(np):
        # calculo da solução aproximada
        xn = xn + dt*f(tn,xn)
        # atualização do valor do tempo
        tn = tn + dt
        
        # anexando os valores calculados nas listas
        x.append(xn)
        t.append(tn)
   
    return [t, x]

#####################################

In [4]:
### Função correspondente ao item b) 
##############################################
# Metodo de Heun (ou Euler melhorado) para o PVI
##############################################

def met_heun( f, t0, tf, x0, np ):
    
    t = [t0]
    x = [x0]
    tn = t0
    x1n = x0
    
    dt = float(tf-t0)/np

    for n in range(np):
        xe = x1n + dt*f(tn, x1n)
        
        x1n = x1n + dt/2*(f(tn, x1n)+f(tn+dt, xe))
        tn = tn + dt
        
        x.append(x1n)
        t.append(tn)
   
    return [t, x]

#####################################

In [22]:
# Função correspondente ao item c) 
##############################################
# Metodo de Runge-Kutta (de ordens superiores) para o PVI
##############################################

def RK4Ordem( f, t0, tf, x0, n ):

  t = [t0]
  x = [x0]
  tn = t0
  x1n = x0  

  dt = float(tf-t0)/n

  for i in range(n):
    
    k1 = dt * f( tn,x1n )
    k2 = dt * f( tn + dt/2 , x1n + (k1/2) )
    k3 = dt * f( tn + dt/2 , x1n + (k2/2) )
    k4 = dt * f( tn + dt , x1n + k3 )

    x1n = x1n + ( 1/6 * (k1 + 2*k2 + 2*k3 +k4))
    tn = tn + dt
    
    x.append(x1n)
    t.append(tn)

  return [t, x]
  
#####################################


## Tarefa #2: Paraquedista em queda livre

Utilize os seus códigos (da tarefa anterior, itens a), b) e c) ) para responder a seguinte questão.   
**Qual será a velocidade vertical atingida  por um paraquedista em queda livre (a velocidade inicial é nula), $25\,s$ após o salto?**


a) Considere que a taxa de variação instantânea da velocidade do paraquedista é dada pela EDO
$$\frac{dv}{dt}=g-\frac{c}{m} v, $$
onde $g=9.8\,m/s^2$ é a aceleração da gravidade, $m=75\,kg$ é a massa do paraquedista e $c=12.5\,kg/s$ é o coeficiente de arrasto correspondente ao paraquedas. 

Para cada método implementado determine um número de passos tal que o erro na aproximação da velocidade final $v(25)$ seja inferior a $10^{-2}$, se sabemos que o valor exato da velocidade  pode ser calculado pela fórmula
$$v(t) = \frac{gm}{c}\left( 1-e^{-(c/m)t}\right).$$

Dica: Isto poderá ser feito por tentativa e erro.

b) Considere uma outra lei para a taxa de variação instantânea da velocidade do paraquedista dada pela EDO
$$\frac{dv}{dt}=g-\frac{C(v)}{m} v, $$
onde a função empiríca $C(v)$ que caracteriza a resistência do paraquedas é dada por 
$$C(v) = c\left[1+\alpha \left(\frac{v}{v_{max}}\right)^\beta\right],$$
em que $c=12.5\,kg/s$, $\alpha=0.2$, $\beta =1.1$ e $v_{max} = 46\,m/s$.  (Observe que no caso $\alpha=0$ este modelo coincide com o apresentado no item a).)

Obtenha a velocidade final usando cada um dos métodos implementados considerando o mesmo número de passos definido no item a)).
 
c) Compare as velocidades obtidas nos itens a) e b). 


In [31]:
## item a) Método de Euler
## velocidade
import math

e = math.e
g = 9.8
c = 12.5
m = 75
passo = 0
## Função da velocidade
def v(t):
  return (g * m / c) * (1 - e ** -( (c/m) * t))

## Define a função que representa a EDO
def f(t,v):
  return g - (c/m) * v

## Testa valores para determinar o número de passos que satisfaz a função
while True:
  passo += 1
  [t, x]= met_euler(f,0,25,0,passo)
  erro = abs(v(t[passo]) - x[passo])
  if (erro < 10**(-2) ): break

print("Velocidade:", x[passo])
print("Passos:", passo)
print("Erro:",erro)
passo = 0

Velocidade: 57.898370544133854
Passos: 790
Erro: 0.009997135755654085


In [32]:
## item a) Método de Heun
## Testa valores para determinar o número de passos que satisfaz a função
while True:
  passo += 1
  [t, x]= met_heun(f,0,25,0,passo)
  erro = abs(v(t[passo]) - x[passo])
  if (erro < 10**-2): 
    break

print("Velocidade:", x[passo])
print("Passos:", passo)
print("Erro:",erro)
passo = 0

Velocidade: 57.87850926962627
Passos: 35
Erro: 0.009864138751979112


In [33]:
## item a) método de Runge-Kutta
## Testa valores para determinar o número de passos que satisfaz a função
while True:
  passo += 1
  [t, x]= RK4Ordem(f,0,25,0,passo)
  erro = abs(v(t[passo]) - x[passo])
  if (erro < 10**-2): break

print("Velocidade:", x[passo])
print("Passos:", passo)
print("Erro:",erro)

passo = 0

Velocidade: 57.88180552525808
Passos: 7
Erro: 0.0065678831201694265


In [44]:
## item b) Método de Euler
## velocidade
import math

e = math.e
g = 9.8
c = 12.5
m = 75
passo = 790

## Função da velocidade
def v(t):
  return (g * m / c) * (1 - e ** -( (c/m) * t))

## Função da resistencia do paraquedas
def C(k):
  coef = 12.5 * ( 1 + (k*0.2/46)**(1.1) ) 
  return coef
## Define a função que representa a EDO
def f(t,v):
  return g - (C(v)/m) * v

## Testa valores para determinar o número de passos que satisfaz a função
[t, x]= met_euler(f,0,25,0,passo)
erro = abs(v(t[passo]) - x[passo])

print("Velocidade:",x[passo])
print("Erro:",erro)
passo = 0

Velocidade: 49.44223533645224
Erro: 8.446138071925958


In [42]:
## item b) Método de Heun
## velocidade
import math

e = math.e
g = 9.8
c = 12.5
m = 75
passo = 35

## Função da velocidade
def v(t):
  return (g * m / c) * (1 - e ** -( (c/m) * t))

## Função da resistencia do paraquedas
def C(k):
  coef = c * ( 1 + (k*0.2/46)**(1.1) ) 
  return coef
## Define a função que representa a EDO
def f(t,v):
  return g - (C(v)/m) * v

## Testa valores para determinar o número de passos que satisfaz a função
[t, x]= met_heun(f,0,25,0,passo)
erro = abs(v(t[passo]) - x[passo])

print("Velocidade:",x[passo])
print("Erro:",erro)
passo = 0

Velocidade: 49.4336902973022
Erro: 8.45468311107605


In [43]:
## item b) Método de Runge-Kutta
## velocidade
import math

e = math.e
g = 9.8
c = 12.5
m = 75
passo = 7

## Função da velocidade
def v(t):
  return (g * m / c) * (1 - e ** -( (c/m) * t))

## Função da resistencia do paraquedas
def C(k):
  coef = 12.5 * ( 1 + (k*0.2/46)**(1.1) ) 
  return coef
## Define a função que representa a EDO
def f(t,v):
  return g - (C(v)/m) * v

## Testa valores para determinar o número de passos que satisfaz a função
[t, x]= RK4Ordem(f,0,25,0,passo)
erro = abs(v(t[passo]) - x[passo])

print("Velocidade:",x[passo])
print("Erro:",erro)
passo = 0

Velocidade: 49.43162827626715
Erro: 8.456745132111102


## Tarefa #3: Dinâmica predador-presa

Para modelar a dinâmica de das populações de duas espécies que interagem: uma como predadora e a outra  como  presa, é usado o seguinte sistema de duas EDOs conhecido como equações de Lotka-Volterra:

$$
\left\{
\begin{aligned}
\frac{d x_1}{dt} &= a x_1-b x_1 x_2\\
\frac{d x_2}{dt} &= -c x_2 + d x_1 x_2
\end{aligned}
\right.
$$

onde $x_1$ e $x_2$ representam o número de presas e predadores, respectivamente, $t$ o tempo e $a$, $b$, $c$, $d$ são coeficientes constantes que caracterizam o crescimento e a interação das populações de presas e predadores.
 
Considere que a dinâmica das populações de alces e lobos de uma região pode ser modelada pelo sistema de Lotka-Volterra com os coeficientes $a=0.3$, $b=0.01111$, $c=0.2106$, $d=0.0002632$. Se inicialmente na região haviam $705$ alces e  $25$ lobos, qual será o tamanho dessas populações  após $20\, anos$?

Utilize o seu código do método de RK de $4^a$ ordem (tarefa #1, item c)) para responder a essa questão. Procure uma aproximação com erro estimado inferior a $0.05$. Explique. 

In [12]:
def RungeK4Ordem( f, t0, tf, x0, n ):
  dt = float(tf-t0)/n

  t = t0
  x = x0

  k1 = dt*f(t0,x0)
  k2 = dt*f(t0+dt/2,x0+k1/2)
  k3 = dt*f(t0+dt/2,x0+k2/2)
  k4 = dt*f(t0+dt,x0+k3)

  x = x + ( 1/6 * (k1 + 2*k2 + 2*k3 +k4))
  t = t + dt

  return x, t

In [47]:
a=0.3
b=0.01111
c=0.2106
d=0.0002632

def f(t,p):
    x = p[0]
    y = p[1]
    dx = a*x - b*x*y
    dy = c*x*y - d*y
    return np.array([dx,dy], float)

t0 = 0
tf = 20 # Crescimento em 20 anos
x0 = np.array([705,25], float) # 705 presas e 25 predadores
N = 1000 # números de passos

predador_prey, t = RungeK4Ordem( f, t0, tf, x0, N )

print("Presas:",predador_prey[0])
print("Predadores:",predador_prey[1])

Presas: 689.7017347905803
Predadores: 394.8098058868576


## Relatório

A explicação do desenvolvimento e implementação das tarefas foi feita no próprio código.

T1 - Implementações

Os códigos de Euler e Heun foram retirados do google colab da aula de PVI. E seguindo as fórmulas do slide de aula, foi implementado a função de runge-kutta.

T2 - Paraquedista

Como demostrado no código referente a implemetação da tarefa, houve uma diferença entre as respostas encontradas e o número de passos exigidos pelas funções. As velocidades encontradas foram bem similares mas a função que apresentou o melhor resultado foi o Runge-kutta com 7 passos. Já os métodos de Heun e Euler apresentaram respectivamente 25 e 790 passos, sendo o método de Euler com o pior resultado, apresentando o maior nº de passos e o maior erro.
Isso deve-se a ordem das implementações, como Runge-Kutta é uma função de quarta ordem e as funções de Heun e Euler tem ordens inferiores, ele possui o melhor resultado entre as implementações.

Na letra b), inserimos agora uma função para calcular a resistência do paraquedas. Assim, possuimos uma diminuição na velocidade do paraquedista durante a queda. Obtemos o seguinte resultado:

Euler: 49.44223533645224

Heun: 49.4336902973022

Runge- Kutta: 49.43162827626715

E como esperado encontramos velocidades e erros similares em todas as funções.


T3 - predador-presa

A função utilizada consiste no método Runge-Kutta de quarta ordem, adaptado para utilizarmos na equação de Lotka-Volterra.
A implementação utilizou os dados oferecidos no enunciado e um N razoavelmente grande (1000 passos) e encontramos uma diminuição das presas e aumento dos predadores. A função funciona da seguinte maneira: quando o número de presas vai diminuindo, a quantidade de predadores aumenta, mas quando existe um aumento considerável de algumas das espécies a outra diminui, e freia o crescimento da outra - E assim por diante.
 